File: packages/totem-sdk/packages/governance/src/proposal.ts
import { createProof } from '@totemsdk/proof'
import type { UnsignedProof } from '@totemsdk/proof'
import type { GovernanceConfig, Proposal, ProposalAction, MembershipSnapshot } from './types.js'
import { computeProposalId } from './ids.js'
import { getMemberWeight } from './snapshot.js'

export function createProposal(params: {
  config: GovernanceConfig
  actions: ProposalAction[]
  title: string
  description: string
  proposer: string
  snapshot: MembershipSnapshot
  createdAt?: number
}): Proposal | string {
  const { config, actions, title, description, proposer, snapshot } = params
  const now = params.createdAt ?? Date.now()

  const proposerWeight = getMemberWeight(snapshot, proposer)
  if (proposerWeight < config.membership.minWeightToPropose) {
    return `proposer weight ${proposerWeight} below minimum ${config.membership.minWeightToPropose}`
  }

  if (!snapshot.entries.has(proposer)) {
    return 'proposer is not a member'
  }

  const proposalId = computeProposalId(config.daoId, proposer, now, actions.length)

  const votingStartsAt = now + config.voting.delayBeforeVotingMs
  const votingEndsAt = votingStartsAt + config.voting.votingPeriodMs

  return {
    id: proposalId,
    daoId: config.daoId,
    title,
    description,
    actions: actions.map((a) => ({ ...a })),
    proposer,
    status: 'draft',
    createdAt: now,
    votingStartsAt,
    votingEndsAt,
    executionDelay: config.voting.executionDelayMs,
    membershipSnapshotHash: snapshot.hash,
  }
}

export function createProposalProofDraft(
  proposal: Proposal,
  issuer: string,
): UnsignedProof {
  return createProof({
    kind: 'custom',
    subject: { id: proposal.id, kind: 'proposal' },
    issuer,
    issuedAt: proposal.createdAt,
    payload: {
      schema: 'totem:governance:proposal/v1',
      proposal: {
        id: proposal.id,
        daoId: proposal.daoId,
        title: proposal.title,
        actionCount: proposal.actions.length,
        proposer: proposal.proposer,
        votingStartsAt: proposal.votingStartsAt,
        votingEndsAt: proposal.votingEndsAt,
        membershipSnapshotHash: proposal.membershipSnapshotHash,
      },
    },
  })
}

export function activateProposal(proposal: Proposal): Proposal | string {
  if (proposal.status !== 'draft') {
    return `cannot activate proposal in status '${proposal.status}'`
  }
  const now = Date.now()
  if (now < proposal.votingStartsAt) {
    return `voting has not started yet (starts at ${proposal.votingStartsAt})`
  }
  return { ...proposal, status: 'active' }
}

export function cancelProposal(proposal: Proposal): Proposal | string {
  if (proposal.status === 'executed' || proposal.status === 'cancelled') {
    return `cannot cancel proposal in status '${proposal.status}'`
  }
  return { ...proposal, status: 'cancelled' }
}

File: packages/totem-sdk/packages/governance/rust/src/proposal.rs
use serde::Deserialize;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::ids::compute_proposal_id;
use crate::snapshot::get_member_weight;
use crate::types::{GovernanceConfig, MembershipSnapshot, Proposal, ProposalAction, ProposalStatus};

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

#[derive(Deserialize)]
pub struct CreateProposalParams {
    pub config: GovernanceConfig,
    pub actions: Vec<ProposalAction>,
    pub title: String,
    pub description: String,
    pub proposer: String,
    pub snapshot: MembershipSnapshot,
    pub created_at: Option<u64>,
}

pub fn create_proposal(params: CreateProposalParams) -> Result<Proposal, String> {
    let now = params.created_at.unwrap_or_else(now_ms);

    if !params.snapshot.entries.contains_key(&params.proposer) {
        return Err("proposer is not a member".to_string());
    }

    let proposer_weight = get_member_weight(&params.snapshot, &params.proposer);
    if proposer_weight < params.config.membership.min_weight_to_propose {
        return Err(format!(
            "proposer weight {} below minimum {}",
            proposer_weight, params.config.membership.min_weight_to_propose
        ));
    }

    let proposal_id = compute_proposal_id(
        &params.config.dao_id,
        &params.proposer,
        now,
        params.actions.len(),
    );

    let voting_starts_at = now + params.config.voting.delay_before_voting_ms;
    let voting_ends_at = voting_starts_at + params.config.voting.voting_period_ms;

    Ok(Proposal {
        id: proposal_id,
        dao_id: params.config.dao_id,
        title: params.title,
        description: params.description,
        actions: params.actions,
        proposer: params.proposer,
        status: ProposalStatus::Draft,
        created_at: now,
        voting_starts_at,
        voting_ends_at,
        execution_delay: params.config.voting.execution_delay_ms,
        executed_at: None,
        execution_tx_id: None,
        vote_tally: None,
        membership_snapshot_hash: params.snapshot.hash,
    })
}

pub fn activate_proposal(proposal: &Proposal) -> Result<Proposal, String> {
    if proposal.status != ProposalStatus::Draft {
        return Err(format!(
            "cannot activate proposal in status '{:?}'",
            proposal.status
        ));
    }
    let now = now_ms();
    if now < proposal.voting_starts_at {
        return Err(format!(
            "voting has not started yet (starts at {})",
            proposal.voting_starts_at
        ));
    }
    let mut p = proposal.clone();
    p.status = ProposalStatus::Active;
    Ok(p)
}

pub fn cancel_proposal(proposal: &Proposal) -> Result<Proposal, String> {
    if proposal.status == ProposalStatus::Executed || proposal.status == ProposalStatus::Cancelled {
        return Err(format!(
            "cannot cancel proposal in status '{:?}'",
            proposal.status
        ));
    }
    let mut p = proposal.clone();
    p.status = ProposalStatus::Cancelled;
    Ok(p)
}
